File: Python/applications/error_estimator.py
import numpy as np

from ..datastructures.applicator import LinearOperatorApplicator
from ..space.triangulation_function import TriangulationFunction
from .heat_equation import HeatEquation


class ErrorEstimator:
    """ An abstract error estimator.

    Arguments:
        g_functional: the functional of the data g.
        u0_functional: the functional of the data u0.
        dirichlet_boundary: whether or not to enforce Dirichlet BC.
    """
    def __init__(self, g_functional, u0_functional, dirichlet_boundary=True):
        self.g_functional = g_functional
        self.u0_functional = u0_functional
        self.dirichlet_boundary = dirichlet_boundary


class AuxiliaryErrorEstimator(ErrorEstimator):
    def __init__(self,
                 g_functional,
                 u0_functional,
                 u0,
                 u0_order,
                 u0_slice_norm,
                 dirichlet_boundary=True):
        super().__init__(g_functional, u0_functional, dirichlet_boundary)
        self.u0 = u0
        self.u0_slice_norm = u0_slice_norm
        self.u0_order = u0_order

    def estimate(
            self,
            heat_dd_d,
            u_dd_d,
            solver_tol=1e-5,
    ):
        Bu_minus_g = heat_dd_d.B.apply(u_dd_d)
        Bu_minus_g -= self.g_functional.eval(heat_dd_d.Y_delta)
        Bu_minus_g_array = Bu_minus_g.to_array()
        A_s = LinearOperatorApplicator(applicator=heat_dd_d.A_s,
                                       input_vec=Bu_minus_g)
        result, _ = A_s.solve(solver='pcg',
                              b=Bu_minus_g_array,
                              M=LinearOperatorApplicator(
                                  applicator=heat_dd_d.P_Y,
                                  input_vec=Bu_minus_g),
                              tol=solver_tol)
        u_dd_d_0 = u_dd_d.slice(i=0,
                                coord=0.0,
                                slice_cls=TriangulationFunction)
        zero_slice_error = u_dd_d_0.error_L2(
            lambda xy: self.u0(xy),
            self.u0_slice_norm,
            self.u0_order,
        )
        terms = (np.sqrt(Bu_minus_g_array.dot(result)), zero_slice_error)
        return np.sqrt(sum([t**2 for t in terms])), terms


class ResidualErrorEstimator(ErrorEstimator):
    @staticmethod
    def mean_zero_basis_transformation(vector):
        """ Transforms the space basis into a mean zero basis."""
        for db_node in reversed(vector.bfs()):
            if db_node.nodes[1].level == 0 or db_node.nodes[
                    1].on_domain_boundary or any(
                        parent.nodes[1].on_domain_boundary
                        for parent in db_node.parents[1]):
                continue
            for parent in db_node.parents[1]:
                db_node.value -= 0.5 * db_node.nodes[1].volume(
                ) / parent.nodes[1].volume() * parent.value

    def estimate(self, u_dd_d, X_d, X_dd, Y_dd, mean_zero=False):
        """ The residual error estimator of Proposition 5.7.

        Arguments:
            u_dd_d: the solution vector.
            X_d: the doubletree `X_delta`.
            X_dd: the doubletree `X_{underscore delta}`.
            Y_dd: the doubletree `Y_{underscore delta}`.
        """
        # First lift u_dd_d onto X_dd.
        u_dd_dd = u_dd_d.deep_copy()
        u_dd_dd.union(X_dd, call_postprocess=None)

        # Create operator/rhs for (X_dd, Y_dd).
        heat_dd_dd = HeatEquation(X_delta=X_dd,
                                  Y_delta=Y_dd,
                                  formulation='schur',
                                  dirichlet_boundary=self.dirichlet_boundary,
                                  use_space_cache=False)
        f_dd_dd = heat_dd_dd.calculate_rhs_vector(self.g_functional,
                                                  self.u0_functional)

        # Calculate the residual wrt X_dd, i.e.  f_dd_dd - S_dd_dd(u_dd_d).
        residual_vector = f_dd_dd
        residual_vector -= heat_dd_dd.mat.apply(u_dd_dd)

        # We mark all the nodes in X_d in X_dd.
        X_d_nodes = X_dd.union(X_d, call_filter=lambda _: False)
        for node in X_d_nodes:
            node.marked = True

        # Return a list of the residual nodes on X_d and X_dd \ X_d.
        res_d = []
        res_dd_min_d = []
        res_dd_d = residual_vector.deep_copy()

        # Do a basis transformation to mean zero space functions.
        if mean_zero:
            self.mean_zero_basis_transformation(res_dd_d)

        def calculate_residual(res_node, other_node):
            # There is nothing to do for metaroots.
            if other_node.is_metaroot(): return
            assert other_node.nodes[0].level >= 0 and other_node.nodes[
                1].level >= 0

            # If this node is in X_d, the residual should be zero,
            # and there is nothing left to do.
            if other_node.marked:
                res_d.append(res_node)
                return
            else:
                res_dd_min_d.append(res_node)

            # This is a node in X_dd \ X_d, we now evaluate the residual
            # using a scaled basis.
            lvl_diff = other_node.nodes[0].level - other_node.nodes[1].level
            res_node.value /= np.sqrt(1.0 + 4**(lvl_diff))

        # Calculate the residual.
        res_dd_d.union(X_dd, call_postprocess=calculate_residual)

        # Validate the boundary conditions.
        heat_dd_dd._validate_boundary_dofs(res_dd_d)

        # Unmark the nodes in X_dd
        for node in X_d_nodes:
            node.marked = False

        return res_dd_d, res_d, res_dd_min_d


class TimeSliceErrorEstimator(ErrorEstimator):
    """ This error estimator measures a time slice error in l2.

    This requires knowledge of the exact solution.
    """
    def __init__(self, u, u_order, u_slice_norm, dirichlet_boundary=True):
        super().__init__(g_functional=None,
                         u0_functional=None,
                         dirichlet_boundary=dirichlet_boundary)
        self.u = u
        self.u_slice_norm = u_slice_norm
        self.u_order = u_order

    def estimate(self, u_delta, times):
        errors = []
        for i, t in enumerate(times):
            sol_slice = u_delta.slice(i=0,
                                      coord=t,
                                      slice_cls=TriangulationFunction)
            errors.append(
                sol_slice.error_L2(
                    lambda xy: self.u[0](t) * self.u[1](xy),
                    self.u_slice_norm(t),
                    self.u_order[1],
                ))

        return np.array(errors)

File: C++/applications/error_estimator.cpp
#include "error_estimator.hpp"

#include "space/integration.hpp"

namespace applications::ErrorEstimator {
namespace {
double u0L2NormSquared(HeatEquation &heat,
                       LinearFormBase<ThreePointWaveletFn> &u0_lf) {
  auto u0_functional = u0_lf.SpaceLF().Functional();
  auto u0 = u0_functional->Function();
  double u0_norm_sq = 0.0;
  auto space_metaroot = heat.vec_X()->Project_1()->node()->vertex();
  assert(space_metaroot->is_metaroot());
  space::Element2D *elem_metaroot =
      space_metaroot->children()[0]->patch[0]->parents()[0];
  assert(elem_metaroot->is_metaroot());
  auto u0_sq = [&u0](double x, double y) { return u0(x, y) * u0(x, y); };
  for (auto &elem : elem_metaroot->children())
    u0_norm_sq += space::Integrate(u0_sq, *elem, 2 * u0_functional->Order());
  return u0_norm_sq;
}

void ApplyMeanZero(
    DoubleTreeVector<ThreePointWaveletFn, HierarchicalBasisFn> *vec) {
  for (auto &dblnode : boost::adaptors::reverse(vec->container())) {
    auto [_, space_node] = dblnode.nodes();
    if (space_node->level() == 0 || space_node->on_domain_boundary()) continue;
    if (std::any_of(space_node->parents().begin(), space_node->parents().end(),
                    [](auto parent) { return parent->on_domain_boundary(); }))
      continue;
    for (auto &parent : dblnode.parents(1)) {
      auto [_, space_parent] = parent->nodes();
      dblnode.set_value(dblnode.value() - 0.5 * space_node->Volume() /
                                              space_parent->Volume() *
                                              parent->value());
    }
  }
}
}  // namespace

GlobalError ComputeGlobalError(const Eigen::VectorXd &g_min_Bu,
                               const Eigen::VectorXd &PY_g_min_Bu,
                               const Eigen::VectorXd &G_u_dd_dd,
                               const Eigen::VectorXd &u0, HeatEquation &heat,
                               const Eigen::VectorXd &u_dd_dd,
                               LinearFormBase<ThreePointWaveletFn> &u0_lf) {
  GlobalError error;
  double error_Yprime_sq = PY_g_min_Bu.dot(g_min_Bu);

  // Compute ||u_0 - u(0)||_L2^2 as ||u_0||^2 - 2<u_0, u(0)> + ||u(0)||^2.
  double u0_norm_sq = u0L2NormSquared(heat, u0_lf);
  double u0_gamma0_u_inp = u0.dot(u_dd_dd);
  double gamma0_u_norm_sq = G_u_dd_dd.dot(u_dd_dd);
  double error_t0_sq = u0_norm_sq - 2 * u0_gamma0_u_inp + gamma0_u_norm_sq;

  error.error = sqrt(error_Yprime_sq + error_t0_sq);
  error.error_Yprime = sqrt(error_Yprime_sq);
  error.error_t0 = sqrt(error_t0_sq);
  return error;
}

double ComputeLocalErrors(
    DoubleTreeVector<ThreePointWaveletFn, HierarchicalBasisFn> *residual_dd_dd,
    bool mean_zero) {
  if (mean_zero) ApplyMeanZero(residual_dd_dd);
  // Do a basis transformation for calculation of the residual wrt Psi_delta.
  double sq_norm = 0.0;
  for (auto &dblnode : residual_dd_dd->container()) {
    if (dblnode.is_metaroot()) continue;
    int lvl_diff = std::get<0>(dblnode.nodes())->level() -
                   std::get<1>(dblnode.nodes())->level();
    dblnode.set_value(dblnode.value() / sqrt(1.0 + pow(4.0, lvl_diff)));
    sq_norm += dblnode.value() * dblnode.value();
  }
  return sqrt(sq_norm);
}

}  // namespace applications::ErrorEstimator
